tests/nn/data_parallel/test_fsdp.py

```diff
@@ -0,0 +1,810 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+import functools
+import itertools
+from math import inf
+import pickle
+import sys
+from typing import Dict
+import unittest
+from unittest import mock
+
+from parameterized import parameterized
+import torch
+from torch import nn
+
+from fairscale.nn.data_parallel import FullyShardedDataParallel
+from fairscale.nn.misc.checkpoint_activations import checkpoint_wrapper
+from fairscale.utils.testing import (
+    DeviceAndTypeCheckModule,
+    DummyProcessGroup,
+    dist_init,
+    get_cycles_per_ms,
+    objects_are_equal,
+    spawn_for_all_world_sizes,
+)
+
+# How to use remote-pdb: https://gist.github.com/sshleifer/9d43351957179c13606e015b072927d4
+# All helper functions called by spawn must be either @classmethod, @staticmethod
+
+_BUFFER_NAME = "vocab_bias"
+
+
+class DistributedTest(unittest.TestCase):
+    def setUp(self):
+        major, minor = torch.__version__.split(".")[:2]
+        major, minor = int(major), int(minor)
+        if major < 1 or (major == 1 and minor < 6):
+            raise unittest.SkipTest("Need pytorch version >= 1.6 due to autocast")
+        if not torch.cuda.is_available():
+            raise unittest.SkipTest("CUDA not available, skipping test")
+        if sys.platform == "win32":
+            raise unittest.SkipTest("NCCL doesn't support Windows, skipping test")
+        if torch.cuda.device_count() < 2:
+            raise unittest.SkipTest("distributed tests require 2+ GPUs, skipping")
+
+    @staticmethod
+    def _train_for_several_steps(model, num_steps, autocast, lr=0.01, norm_type=None):
+        model_device = next(model.parameters()).device
+        # use SGD with momentum instead of Adam, since Adam is scale invariant
+        # and this makes it bad for tests
+        optim = torch.optim.SGD(model.parameters(), lr=lr, momentum=0.9)
+        for _ in range(num_steps):
+            optim.zero_grad()
+            with torch.cuda.amp.autocast(enabled=autocast):
```

Comment:
Had a quick sneak at the PR, was wondering if this test will pass if we add a `GradScaler` object? If weights/states are sharded across will we need to use something similar to the `ShardedGradScaler`: https://github.com/facebookresearch/fairscale/blob/master/fairscale/optim/grad_scaler.py#L24

Reply:
Good question and I think you're probably right. PR's welcome :)